### ISSUE TITLE
Trying to understand behavior of rotationX, rotationY, rotationZ

### ISSUE THREAD

**Issue by aatishb** [processing/p5.js#1249]
I've been testing out the `rotationX`, `rotationY`, `rotationZ` variables. From what I understand, these are the same as `beta`, `gamma`, `alpha` [Tait-Bryan](https://dev.opera.com/articles/w3c-device-orientation-usage/#coordinatesystem) angles provided by the Device Orientation API. 

I think they work very well provided the rotation stays under 90 degrees in any direction. However, once the rotation angle hits 90 degrees I'm running into some unexpected behavior. I did these tests on a [graphing sketch](https://stc209.com/aatishb/graph_device_rotation/) I put together (it just plots values of `rotationX`, `rotationY`, `rotationZ`).

The Z rotation works as expected. Here I'm rotating at a steady rate around Z axis in one direction, and then in the reverse direction.
<img src="https://cloud.githubusercontent.com/assets/1878638/12981771/18ee5fae-d0b0-11e5-9ef3-b9dbfbb23605.jpg" alt="Z rotation" width="300">

However, I'm running into some weirdness with X and Y rotations. Here I'm rotating at a steady rate around Y axis in one direction, and then in the reverse direction.
<img src="https://cloud.githubusercontent.com/assets/1878638/12981922/c9197a12-d0b0-11e5-9db3-f097cd8f1622.jpg" alt="Y rotation" width="300">
The Y rotation goes from -90 to 90, which I think is the correct behavior (see [this issue](https://github.com/processing/p5.js/issues/1243)). However, notice that when the Y rotation hits 90 or -90, the X and Z rotation values jump discontinuously. I suspect this might be the 'gimbal lock' phenomenon. ([ref 1](http://w3c.github.io/deviceorientation/spec-source-orientation.html#ref-gimballock), [ref 2](https://dev.opera.com/articles/w3c-device-orientation-usage/#eulerlimitations), [ref 3](http://danceswithcode.net/engineeringnotes/rotations_in_3d/rotations_in_3d_part1.html))

Next, I'm rotating at a steady rate around X axis in one direction, and then in the reverse direction.
<img src="https://cloud.githubusercontent.com/assets/1878638/12981899/b91f8bce-d0b0-11e5-8795-9e0fd9a8f9f8.jpg" alt="X rotation" width="300">
What's strange here is that when X rotation is +90 or -90, I see discontinuities in the Y and Z rotation (as well as a gap in the X rotation). I don't understand why this happens.

You can see these discontinuities with this [box rotation app](https://stc209.com/aatishb/box_rotation/) as well.

So my questions are: 
1. Are these the expected behaviors for these variables? 
2. Do other folks run into the same issues?  (I saw these in Safari and Chrome on iPhone IOS 9)

I wonder if it's worth trying to find a way to use these variables to get a consistent device orientation that works smoothly without discontinuities (I can try to help with this). This may involve getting the rotation matrix from the alpha, beta, gamma variables and then working with that... or there might be a simpler way to do it. On the other hand, it might be overkill for most applications. Thoughts?

**Comment by limzykenneth** [comment]
@aatishb Back when I was implementing these for #984 and #909, I did noticed these kind of behaviour. I couldn't solve that problem because it seemed to me that it is a limitation on the hardware (a phone held vertically or face down have very erratic behaviour). I think if we can somehow abstract away the direct relationship between rotationX/Y/Z to e.alpha/beta/gamma then we might be able to solve this.

I plan on working on #1243 during the weekend and if you like we can work on it together. I'll especially need help on gimbal lock if that really is the problem in the first case as I'm not that familiar with it.

**Comment by aatishb** [comment]
@limzykenneth Agree, and I'll try to help.

I did some more experimenting, and I think I now understand this better. It seems that the discontinuities in the graphs above are artifacts of the way the coordinate system is set up, and they shouldn't have any physical effect on the actual orientation of an object.

With the Y rotation issue (second graph above), it boils down to this: Rotating by theta1, theta2, -90 around Z,X,Y axes gives the same orientation as rotating by 180+theta, 180-theta2, 90. 

You can test this by seeing that you get the same result when you replace the highlighted part with the commented code below:

![clip1](https://cloud.githubusercontent.com/assets/1878638/12996620/75f632b8-d0ff-11e5-8768-afbdaa605671.png)

The reason I was running into these discontinuities earlier is that I was executing the rotations in the wrong order. It's important that you do the rotations in the order Z,X,Y (the convention used by the [deviceOrientation API](http://w3c.github.io/deviceorientation/spec-source-orientation.html)). When I make this change to the [box rotation app](https://stc209.com/aatishb/box_rotation/) it seems to work smoothly for all angles and orientations.

Similarly, for the X rotation issue (third graph above), I think it comes down to this: rotating by theta1, 90, theta2 around Z,X,Y axes gives the same orientation as rotating by theta1+x, 90, theta2-x. 

Again, you can test this by replacing the highlighted code with the commented code below: 

![clip2](https://cloud.githubusercontent.com/assets/1878638/12996623/7adf66dc-d0ff-11e5-9dee-7aa81702422a.png)

So to summarize, I think the discontinuities in the graph above shouldn't affect real rotations, **provided you execute the rotations in the correct order**. If you change the order of rotations, all bets are off! Might be worth mentioning in the documentation for these rotation variables!

**Comment by limzykenneth** [comment]
@aatishb I'll be testing everything tomorrow and make changes if necessary (docs as well). You can fork the repo and push any of the changes you made to my fork [here](https://github.com/limzykenneth/p5.js), I can submit a PR from my side when we're done.

**Comment by limzykenneth** [comment]
@aatishb I'm a bit confused now (mostly by myself :P). The changing in order definitely fixed the box rotation and i think the graph app as well I just don't really understand why, something to do with matrix maths I guess? The more confusing part is that although the implementation of `deviceturned()` is assuming `rotationY/e.gamma` is +/-180, it somehow still works for me....you can try it [here](http://sick-juice.surge.sh), the logged value for `rotationY` is going +/-90 but the `deviceturned()` function still works as intended. Blind luck??

**Comment by aatishb** [comment]
@limzykenneth Yeah, 3D rotations are kind of mind-bending. It's easy to demonstrate that the order matters though. Take your phone and put it face up on a table. Now do a series of 90 degrees rotations around any ordering of X, Y and Z. Now start again, but with a different ordering. You'll see that the final orientation of your phone depends on the order of axes you chose.

Or try this:
![boxrot](https://cloud.githubusercontent.com/assets/1878638/13028004/f757c3a4-d22f-11e5-865f-13c29d07c00f.png)

BTW, were you able to get rid of the discontinuities in the graphing app? If so, I'm interested how. 

I'll try out your deviceturned() app now.

**Comment by aatishb** [comment]
Hmm.. the deviceturned app doesn't work so well for me with Y turns. It's a little erratic in registering them. Most of the times it just increments the X and the Z instead. (which agrees with the graphs above). I think this might be why deviceTurned() seems to work ok.. when you turn on Y it probably registers the 'virtual' X and Y rotations (see 2nd graph) as valid turns. So checking the specific (turnAxis == whatever) gives you the wrong values on a Y turn.

**Comment by aatishb** [comment]
I think the crux of the problem is: there are multiple combinations of angles that correspond to the same physical orientation (this is a problem working with Euler angles).

So e.g. when the deviceOrientation angle variables reach a Y angle of 90 degrees, the representation flips between (thetaX, -90, thetaZ) and (180-thetaX, 90, 180+thetaZ). I think this doesn't affect 3d rotations of objects because if you execute these rotations in the order Z->X->Y, you get the same final orientation. So the box app is unaffected. But if you're *just* using thetaX, thetaY, or thetaZ, you'll see spurious rotations where there aren't any real ones. 

I see two potential ways out: 

1. introduce a bunch of if statements keeping track of X, Y and Z angles to determine which 'real' rotation has taken place. e.g. if X and Z all just changes by 180 degrees and Y just flipped from -90 to 90, we interpret this as a Y turn, not a X or Z turn. This will be a bit messy but might possibly work (and would be fast if it does).

2. the harder route is to use a different representation that isn't susceptible to this problem. The way people usually do this is to represent orientations using either quaternions or a rotation matrix. I could look into how doable this is, but that's a slightly longer project to work out.

**Comment by limzykenneth** [comment]
@aatishb I think I understand the 3-D rotation thing now.

On closer look the graphing app is still breaking, maybe i'm just moving it too slowly for the effect to be pronounced. I'll look into it.

![img_3683](https://cloud.githubusercontent.com/assets/7543950/13028516/1397d822-d269-11e5-8157-106cbf12b899.PNG)


If I remember correctly, back when these were amended by me, there was a discussion about whether to separate the `deviceTurned()` function into three function ie `deviceTurnedX/Y/Z()`, the reason we didn't go with that is not to bloat the whole library too much and add unneeded complexity for the user.

For solution 1, would it be unintuitive if someone actually does something like a 180 X followed by 180 Z but end up with a Y turn?

I'm gravitating towards solution 2 but I'm not familiar with that area, if you are up for it you can take your time trying to implement that. In the meantime, I'll see if I can implement something like solution 1.

I'll make a quick PR to fix the docs for `rotationY` first.

**Comment by aatishb** [comment]
@limzykenneth For solution 1, I was thinking about comparing current rotations to rotations in the previous frame. If some did a 180 X or 180 Z, unless they did this very, very quickly, they wouldn't get a 180 degree jump in a single frame.. the angle would continuously vary between 0 to 180. But in the situation above, when Y approaches +/- 90, the jump is sudden and discontinuous.

The X rotation issue is [more subtle](https://github.com/processing/p5.js/issues/1249#issuecomment-183147843). But when I tested it out deviceTurned() seemed to be working ok for X rotations so maybe it's not worth getting to deep into these quirks.

I'll work on implementing solution 2 as a new rotation function.. something that converts the alpha, beta, gamma variables to an alternate representation that doesn't have these ambiguities (quaternions or rotation matrix). Then we can call this new representation to see if a turn happens, and that should do away with the need for these if statements. I think I can make progress on this within a week or two.

**Comment by aatishb** [comment]
One suggestion. I'd like to add this simple [box rotation example](https://stc209.com/aatishb/box_rotation/sketch.js) in the documentation for the rotationX, rotationY, rotationZ variables, and a few lines to the documentation explaining how the rotation order matters. Could you either incorporate that, or help me understand how I could do it? (I'm new to contributing and not totally sure how things work.) Thanks!

**Comment by limzykenneth** [comment]
@aatishb That's great, thanks.

I think the box rotation example will be a great addition for learning how these work. For documentation example, you can follow instructions [here](https://github.com/processing/p5.js/wiki/Inline-documentation) while for examples to go on the examples page, follow instructions [here](https://github.com/processing/p5.js-website/wiki/Adding-examples), or if you prefer to just work on solution 2, I can handle adding these for you, just say the word. @lmccart do you think the example should go with the docs or as an example on its own?

**Comment by lmccart** [comment]
hey thanks for the in-depth look into this both of you! i think the box rotation example looks simple enough to include right in the docs.

**Comment by aatishb** [comment]
@lmccart happy to help!

@limzykenneth Thanks for the pointers. So when the documentation in the code is edited and pushed to the github, that automatically generates the reference, with then goes to the p5 website? Does it work the other way too, if you edit a page on the p5 website and push the change, is the reference updated in the source code? I'm finding the connection between the code and online documentation a little confusing.

It'd be great if you can handle these changes, thanks! I can go through the reference when you're done (once I wrap my head around how to do so) and if I have any edits I can apply them afterwards. And I'll work on solution 2.

**Comment by limzykenneth** [comment]
@aatishb The documentation itself ie. anything under http://p5js.org/reference/ is generated form the inline docs using YUIDoc. They will periodically get pushed to p5.js-website repo and updated on p5js.org. The generated file is https://github.com/processing/p5.js-website/blob/master/reference/data.json which is eventually fed through Backbone which renders the view. It doesn't work the other way though.

**Comment by lmccart** [comment]
examples look great, thanks for your work on this! is this issue good to close or is there more here to be figure out still?

**Comment by aatishb** [comment]
@lmccart The documentation is fixed. deviceTurned() also works pretty well, the remaining issues are:

1. deviceTurned() reports the turnaxis incorrectly on y rotations from 0 to +/- 90 
2. (related) the individual rotation variables can unexpectedly change when approaching +/-90 on X or Y axis (due to non-unique Euler angles representing a given orientation)

I'm working on a solution for 2, within the next week or so. I think @limzykenneth was working on a solution for 1 (but now I'm wondering if it makes sense to hold off on this because solving 2 should also solve 1.. @limzykenneth what do you think?)

**Comment by limzykenneth** [comment]
A quick hack for solution one would just be changing some numbers for the range conversion, I'll try to do that if I can find some spare time but I think solving the bigger problem that we had should remove the need for these hack in the first place.

**Comment by aatishb** [comment]
@lmccart @limzykenneth Sounds good, I agree, let's close this. I'll try to work on the remaining issue soon.

**Comment by lmccart** [comment]
sounds good!

**Comment by aatishb** [comment]
hi @limzykenneth, I made progress with implementing rotations more robustly, by adding quaternion & rotation matrix support. You can find my fork [here](https://github.com/aatishb/p5.js).

Some examples:
1. [Box Rotation](https://stc209.com/aatishb/formobile/box_rotation/): rotates two boxes at the same positions using rotationX/Y/Z variables, rotation matrix, and quaternions. Since all 3 methods agree, you should (hopefully) only see one pair of boxes.
2. [Device Turned](https://stc209.com/aatishb/formobile/quaternion_print_rotation/) implementation using quaternions. This should resolve the issues we ran into above when using rotationX/Y/Z variables separately.

Could you test it out and let me know if it works ok? Appreciate any feedback.

**Comment by limzykenneth** [comment]
@aatishb I'll check it out today or this weekend. Great job, thanks!

**Comment by limzykenneth** [comment]
The box rotation app works perfectly while the device turn app I still need to do some test on it. In the mean time could @indefinit have a look at this as well, there's some functions added to the 3d matrix transforms so I think it'll be best if you can have a look and make sure there's no major conflict with the work being done on the 3D part.

**Comment by aatishb** [comment]
@limzykenneth Sounds good. [Here's] (https://stc209.com/aatishb/formobile/quaternion_device_turned/) a slightly more succinct version of the device turn app. You can change the line (quatAngleBetween >= 90) to change the threshold angle. The basic idea is to compare quaternions (which is just a fancy way to store the 3D orientation of an x,y,z frame) to identify when a turn has occurred, and in what direction. This should get around all the issues we ran into with using individual angles (which will always have singularities). 

@indefinit - the main additions I made to the matrix 3d section:
1. added rotateMatrix() and rotateQuaternions(), which allow you to use a 3x3 rotation matrix or a quaternion as input to control rotations. Also added the wrapper functions for these (following the way that rotate() is called)
2. added some helper functions to allow you to create and manipulate the quaternions and rotation matrices... (I'm not sure if the matrix 3d section was the right place to put these.)

If these additions make sense, I can continue to develop these (still need to add documentation, and do some more tests to make sure it all works). Look forward to hearing your feedback/thoughts.

**Comment by limzykenneth** [comment]
@aatishb I had a closer looked through the changes and there are a couple things:

1. Is it possible to have the quaternion functions not exposed and it does the job in the background instead. It might introduce too much complexity for beginners. It would be best if it could mirror Processing at least in terms of the 3D rotations.

2. It seems deviceTurned is set to trigger whenever the quaternion angle is 90 which is alright, the problem is with

```
background(255);
  
  if(frameCount < 5){
   w0 = quaternion(rotationX,rotationY,rotationZ); //stores quaternion for originial orientation
  }

  var w = quaternion(rotationX,rotationY,rotationZ); // quaternion for current orientation

  // tells us if the device is turned by 90 degrees
  deviceTurned(w,w0)
```

`deviceTurned()` was intended to be triggered kind of like an event when the device was turned 90 degrees so the user don't have to pass in arguments and just use it out of the box. That will also mean that the "original position" quaternion will need to be updated whenever the rotation direction changes.

I think if you can move the quaternion initialization for rotationX/Y/Z (ie `w0 = quaternion(rotationX,rotationY,rotationZ);`) into the source area of `deviceTurned()` along with the deviceTurned function you have in the sketch, which will be another function to be called internally, that will be perfect.

**Comment by aatishb** [comment]
Hi @limzykenneth, sure, I can work on making those changes. 

I do feel like being able to use quaternions or rotation matrices might be a useful addition to p5, particularly for people working with device orientation, but I can see that it might be too much for beginners. If there's interest down the road, I'm happy to work on making these more intuitive and easy to use. cc @lmccart

**Comment by JMandel1027** [comment]
Thank you @aatishb @limzykenneth for this work! really really great, this has helped me a ton for an interactive sound walk I'm implementing with p5.sound. :)

**Comment by aatishb** [comment]
@JMandel1027 would love to hear what you ended up using.
@lmccart @limzykenneth Just checking whether there's any interest in incorporating quaternions into p5? I have a fork with quaternions implemented and some test sketches written, it seems to work in all my tests. If there's interest I can try to incorporate these to the main branch. No worries if not.

**Comment by jtoy** [comment]
@aatishb and @lmccart why is quaternion support not directly in p5js? @aatishb anyway to get your changes into main? I check every year or so and see that the 3D stuff is not really a priority for p5. If there is interest I would be interested in adding it in.
